text/0000-oci-image-annotations.md

```diff
@@ -0,0 +1,115 @@
+
+# Definitions
+[definitions]: #definitions
+
+[Buildpacks](https://buildpacks.io/docs/for-app-developers/concepts/buildpack/) are [OCI images](https://github.com/opencontainers/image-spec/blob/v1.1.0/README.md). [Annotations](https://github.com/opencontainers/image-spec/blob/v1.1.0/annotations.md) are optional properties that can be applies to image manifests and descriptions providing mechanism to communicate metdata. The [Pre-Defined Annotation Keys](https://github.com/opencontainers/image-spec/blob/v1.1.0/annotations.md#pre-defined-annotation-keys) are a standardized set of annotations that can be used to convey metadata in a consistent way between image authors and users.
+
+
+
+# Motivation
+[motivation]: #motivation
+
+Knowing the origin and other metadata for a buildpack (which is an OCI image) is very helpful. Some examples of such use cases include finding release notes, user manuals, bug reporting procedures, and license information. Currently, it can be difficult to find the source control repository of a buildpack as that information is not available in a standard way.
+
+The OCI Image Format Specification's Pre-Defined Annotation Keys provide a standardized way to discover additional information about an OCI image. Because these annotations are standardized and widely used, tools have come to use them. For example, [Snyk](https://snyk.io/blog/how-and-when-to-use-docker-labels-oci-container-annotations/) and [Renovate](https://github.com/renovatebot/renovate/blob/34.115.1/lib/modules/datasource/docker/readme.md use these annotations.
+
+The outcome will be that users and tools will be able to gather more information about buildpacks, faciliating use cases such as gathering releases notes and finding documentation.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The `buildpack` tool will set the following OCI annotations on the images it produces:
+* `org.opencontainers.image.source`
+* `org.opencontainers.image.revision`
+* `org.opencontainers.image.title`
+* `org.opencontainers.image.version`
+* `org.opencontainers.image.url` (when possible)
+* `org.opencontainers.image.description` (when possible)
+
+The target personas as buildpack users, platform operators, and platform implementors. Any of those groups will be able to more easily understand the origin (source), version, and other information about the buildpack. This information can then be used manually or with the aid of tools to get release notes which aid these personas in making informed decisions.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When building the buildpack, the `buidpack` tool can get the values for the `org.opencontainers.image.source` and `org.opencontainers.image.revision` annotations from git. `org.opencontainers.image.source` is derived from the git origin and `org.opencontainers.image.revision` is the git commit ref.
+
+The other annotation values come from `buildpack.toml` mapped to OCI annotations as follows:
```

Comment:
What happens when an app has more than one buildpack? The proposed schema doesn't support more than one entry.

Reply:
The RFC is for the buildpacks themselves - not for the apps built by the buildpacks.

For example, it would apply to the [Paketo Java buildpack, `gcr.io/paketo-buildpacks/java`](https://github.com/paketo-buildpacks/java), not to the application it builds.

Reply:
Ah ok. I think the RFC wording might need adjusting to make that clearer.

For example here:

> The `buildpack` tool should set OCI annotations on the OCI artifacts it produces

"buildpack tool" isn't a phrase that's really used by this project, and read to me as meaning "the buildpack".

So it then read as:

"The buildpack should set OCI annotations on the OCI artifacts it produces"

...and the only OCI artifacts created when the buildpack runs is the app OCI image.

It sounds like what you're really proposing is a change to the buildpack distribution specification:
https://github.com/buildpacks/spec/blob/main/distribution.md

Reply:
> It sounds like what you're really proposing is a change to the buildpack distribution specification:

Combined with proposing that the Pack CLI implement these changes, so that when a buildpack maintainer runs `pack buildpack package` to package their buildpack into the OCI image form of distribution it has the proposed labels set.

Reply:
Instead of:
> The `buildpack` tool should set...

I've updated the RFC to say:
> `pack buildpack package` should set

Does that eliminate the confusion?

> It sounds like what you're really proposing is a change to the buildpack distribution specification:
> https://github.com/buildpacks/spec/blob/main/distribution.md

I believe so - is there something I should do in this RFC to indicate as much

Reply:
> Does that eliminate the confusion?

Yeah that does thank you.

> I believe so - is there something I should do in this RFC to indicate as much?

So I think this comes down to whether this RFC is proposing only a change to the Pack CLI's own implementation, or a change to the specification which then the Pack CLI (plus any other implementations) would implement. If the latter, than the spec change could be described under the "spec change" section of the RFC template.

Reply:
I've updated all occurrences of "`buildpack` tool" to be "`pack` tool" for consistency